File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseKGroup(head *ListNode, k int) *ListNode {
    dummyNode := &ListNode{0, head}

    prevGroupEnd := dummyNode

    for {
        curGroupEnd := prevGroupEnd

        for i := 0; i < k; i++ {
            curGroupEnd = curGroupEnd.Next
            if curGroupEnd == nil {
                return dummyNode.Next
            }
        }

        nextGroupStart := curGroupEnd.Next
        curGroupStart := prevGroupEnd.Next

        var prev *ListNode
        cur := curGroupStart

        for cur != nextGroupStart {
            nxt := cur.Next
            cur.Next = prev
            prev = cur
            cur = nxt
        }

        prevGroupEnd.Next = prev

        curGroupStart.Next = nextGroupStart

        prevGroupEnd = curGroupStart
    }

}

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummyNode = ListNode()

        dummyNode.next = head

        prev_group_end = dummyNode

        while True:
            cur_group = prev_group_end
            for _ in range(k):
                cur_group = cur_group.next

                if not cur_group:
                    return dummyNode.next
                
            
            next_group_start = cur_group.next
            group_start = prev_group_end.next

            prev, cur = None, group_start

            while cur is not next_group_start:
                nxt = cur.next
                cur.next = prev
                prev = cur
                cur = nxt
            
            prev_group_end.next = prev
            group_start.next = next_group_start


            prev_group_end = group_start